CppSamples/DisplayInformation/ShowGrid/ShowGrid.cpp

```diff
@@ -0,0 +1,248 @@
+
+  m_grid = new LatitudeLongitudeGrid(this);
+}
+
+ShowGrid::~ShowGrid() = default;
+
+void ShowGrid::init()
+{
+  // Register classes for QML
+  qmlRegisterType<MapQuickView>("Esri.Samples", 1, 0, "MapView");
+  qmlRegisterType<SceneQuickView>("Esri.Samples", 1, 0, "SceneView");
+  qmlRegisterType<ShowGrid>("Esri.Samples", 1, 0, "ShowGridSample");
+}
+
+MapQuickView* ShowGrid::mapView() const
+{
+  return m_mapView;
+}
+
+// Set the view (created in QML)
+void ShowGrid::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  m_mapView->setGrid(m_grid);
+
+  emit viewChanged();
+}
+
+SceneQuickView* ShowGrid::sceneView() const
+{
+  return m_sceneView;
+}
+
+// Set the view (created in QML)
+void ShowGrid::setSceneView(SceneQuickView* sceneView)
+{
+  if (!sceneView || sceneView == m_sceneView)
+    return;
+
+  m_sceneView = sceneView;
+  m_sceneView->setArcGISScene(m_scene);
+
+  // Set the SceneView initially to be invisible
+  m_sceneView->setVisible(false);
+
+  emit viewChanged();
+}
+
+
+void ShowGrid::setViewType(const QString& viewType)
+{
+  m_currentViewType = viewType;
+  if (viewType == "MapView")
+  {
+    m_mapView->setViewpointAsync(m_sceneView->currentViewpoint(ViewpointType::BoundingGeometry), 0);
+
+    m_mapView->setVisible(true);
+    m_sceneView->setVisible(false);
+
+    m_sceneView->setGrid(nullptr);
+    m_mapView->setGrid(m_grid);
+  }
+  else if (viewType == "SceneView")
+  {
+    m_sceneView->setViewpointAsync(m_mapView->currentViewpoint(ViewpointType::BoundingGeometry), 0);
+
+    m_mapView->setVisible(false);
+    m_sceneView->setVisible(true);
+
+    m_mapView->setGrid(nullptr);
+    m_sceneView->setGrid(m_grid);
+  }
+
+  emit propertiesChanged();
+}
+
+void ShowGrid::setGridType(const QString& gridType)
+{
+  if (m_grid)
+  {
+    delete m_grid;
+    m_grid = nullptr;
+  }
+
+  m_currentGridType = gridType;
+
+  // Create a new Grid of the selected type
+  if (gridType == "LatLong")
+    m_grid = new LatitudeLongitudeGrid(this);
+  else if (gridType == "MGRS")
+    m_grid = new MGRSGrid(this);
+  else if (gridType == "UTM")
+    m_grid = new UTMGrid(this);
+  else if (gridType == "USNG")
+    m_grid = new USNGGrid(this);
+
+  // Set the grid on the current view
+  if (m_mapView->isVisible())
+    m_mapView->setGrid(m_grid);
+  else
+    m_sceneView->setGrid(m_grid);
+
+  // Set properties from current UI values
+  setLineColor(m_currentLineColor);
+  setLabelColor(m_currentLabelColor);
+  setLabelPosition(m_currentLabelPosition);
+  setLabelFormat(m_currentLabelFormat);
+
+  emit propertiesChanged();
+}
+
+void ShowGrid::setGridVisible(bool visible)
+{
+  m_grid->setVisible(visible);
+}
+
+void ShowGrid::setLabelsVisible(bool visible)
+{
+  m_grid->setLabelsVisible(visible);
+}
+
+void ShowGrid::setLineColor(const QString& lineColor)
+{
```

Comment:
Not quite for all of them, because when we create a new grid, we apply all the current UI settings to it. So if the user has changed the grid to have blue lines and then we swap from a UTM grid to a USNG grid, we use the `setLineColor` method to overwrite the default grid properties and we want that to fire even if `m_lineColor == lineColor`, where m_lineColor is still reflecting what we have in the UI and lineColor is being supplied from the UI. I can check things like m_grid->isVisible() though because that's just a boolean, but things like colors and placement enums are just a bit arduous to check and will bloat the code. If I remember correctly, setting a property to itself is a no-op in RTC so an initial check wouldn't save much time either.

```cpp
  // Set properties from current UI values
  setGridVisible(m_gridVisible);
  setLabelsVisible(m_labelsVisible);
  setLineColor(m_currentLineColor);
  setLabelColor(m_currentLabelColor);
  setLabelPosition(m_currentLabelPosition);
  setLabelFormat(m_currentLabelFormat);
  ```